platform/engineering/design-docs/2020-04-16-integration-testing-with-consumer-driven-contract-tests.md

```diff
@@ -0,0 +1,304 @@
+
+   The assumption henceforth is that the CI workflow is done in Circle.
+3. **The CI job invokes a script to publish the pacts to a broker.** The command to publish should require a version to label the consumer and optionally accept tags. The published pacts should be versioned to the current commit hash of `vets-website`. For builds to be deployed, the pacts should also be tagged with the appropriate environment. *Tagging is important to ensure that we verify only the pact associated with the build that is getting deployed and not pacts generated from other branches.*
+
+   This assumes that the broker is accessible from CI. There may be some work involved in configuring the the broker to allow access depending on where it's hosted.
+4. Publishing the pacts should trigger a webhook in the broker that invokes a Rake task in `vets-api`:
+   ```
+   bundle exec rake pact:verify
+   ```
+   This task pulls all relevant pacts from the broker and verifies that the expected responses match the actual responses when running the requests through a running instance of the API. For any states specified in the pacts, the API will set up any matching states already defined in `provider_states.rb`. If there are no matching states, the verification will fail.
+   
+   [Pact Broker webhooks can be created using the CLI](https://github.com/pact-foundation/pact_broker-client/#create-webhook), whether it's from the Ruby gem or NPM package. **The webhook should just check for `contract_content_changed` because the broker will automatically pass any unmodified contracts that previously passed verification."**
+   
+   While the pacts are being verified, the `vets-website` build can proceed to E2E tests.
+
+   **If the pacts are coming from a feature branch and the API has not been updated to match the new expectations, verification should be expected to fail.** In that case, once the pacts have been published, the BE developers working on updating the API can run the following task during development to verify their changes:
+   ```
+   bundle exec rake pact:verify:at[http://build-box/MyConsumerBuild/latestSuccessful/artifacts/my_consumer-my_provider.json]
+   ```
+   The actual URL should point to the contract that was published from the `vets-website` feature branch. After the `vets-api` feature branch is merged to master, that pact should be able to pass verification. 
+5. **The verification results are published back to the broker** with the version of the API that verified the pacts. Again, we can use the commit hash to version, but this time it's from `vets-api`.
+   The broker maintains a matrix matching the consumer version and provider version.
+6. **The `vets-website` build runs a check to see if it can be deployed.**
+   ```
+   pact-broker can-i-deploy --broker-base-url=BROKER_BASE_URL
+   ```
+   This check is dependent on the published verification results.
```

Comment:
what happens if the verification task fails and doesn't post results back?

Reply:
The `can-i-deploy` check would fail since there's no passing status on verification, so the code wouldn't be mergable/deployable until the verification task passes. In Circle, we can re-run specific steps, so if anything weird is going on in that verification step, we can resolve any issues with the task and re-run it.